Documentation/Components/Inter-Component Communication.md

```diff
@@ -0,0 +1,410 @@
+
+# Inter-Component Communication
+
+
+
+
+## Summary
+
+- The APIs described in this document are used to access, from within a handler's `handle()` function, the functionality provided by another handler
+- The `InvocableHandler` protocol, which interits from `IdentifiableHandler`, is used to denote `Handler`s which can be remotely invoked (ie, from within a `Handler.handle()` function)
+- The `RemoteHandlerInvocationManager` type implements a helper object which is responsible for coordinating inter-component communication
+- This helper object (you can think of this as a kind of mediator) has knowledge of the web service's structure and use that to determine how to realise interactions between handlers
+- The API as currenly defined only supports invoking components which use the request-response communication pattern
+
+
+
+## The `InvocableHandler` protocol
+
+```swift
+public protocol InvocableHandler: IdentifiableHandler where Response: Decodable {
+    typealias ParametersStorageProtocol = InvocableHandlerParametersStorageProtocol
+    associatedtype ParametersStorage: ParametersStorageProtocol = _EmptyParametersStorage<Self> where ParametersStorage.HandlerType == Self
+}
+```
+
+An `InvocableHandler` is an `IdentifiableHandler` whose `Response` type conforms to `Decodable`, and which optionally defines a custom type for passing parameters to a remote invcation (see below for an explanation of how parameter passsing works).
+
+
+## Defining and using invocable handlers
+
+Simple example of a handler which accesses another handler's functionality.
+
+```swift
+struct RandomNumberGenerator: InvocableHandler {
+    class HandlerIdentifier: ScopedHandlerIdentifier<RandomNumberGenerator> {
+        static let main = HandlerIdentifier("main")
+    }
+    let handlerId = HandlerIdentifier.main
+    
+    @Parameter var lowerBound: Int
+    @Parameter var upperBound: Int
+    
+    func handle() -> Int {
+        Int.ramdom(in: lowerBound..<upperBound)
+    }
+}
+
+
+struct Greeter: Handler {
+    private var RHI = RemoteHandlerInvocationManager()
+    
+    @Parameter var name: String
+    
+    func handle() -> EventLoopFuture<String> {
+        RHI.invoke(
+            RandomNumberGenerator.self,
+            identifiedBy: .main,
+            parameters: [
+                .init(\.$lowerBound, 0),
+                .init(\.$upperBound, name.length)
+            ]
+        ).map { number in
+            "Hello, \(name)! Your lucky number is \(number)."
+        }
+    }
+}
+```
+
+
+
+## The Remote Invocation Manager
+
+
+The invocation manager acts as a mediator overseeing all inter-handler interactions for a handler. Each handler which wants to other handlers defines its own instance of the invocation manager.
+
+
+### What's an interaction?
+
+An interaction is an event where date are transferred between two handlers, and which was initiated within one of the two handler's `handle()` functions.
```

Comment:
```suggestion
An interaction is an event where data is transferred between two handlers, and which was initiated within one of the two handler's `handle()` functions.
```

I would use "data is" but "data are" is apparently also correct: https://www.theguardian.com/news/datablog/2010/jul/16/data-plural-singular